### DISCUSSION TITLE
Api server contribution

### DISCUSSION THREAD

**Discussion by TheDropZone** [browser-use/browser-use#8] [Ideas]
Love the project, the direction, and the layout of everything! I'd love to be able to use a tool like this via an api call, and figured that could be a potential contribution to the project? Do you all have any pointers or recommendations on layout or design for something like that?

Current thoughts are:
- FastAPI
- model a rest api around the try.py cli example
- provide configuration for step count, and return data from the run in json format
- stretch: place inside a docker container for easy api deployment

Also, are there any contribution guidelines for this project? 

Thanks!

**Comment by gregpr07** [comment]
hey, @TheDropZone sorry didn't realise there are Discussion.. :)

Would you be interested in a API template with FastAPI + docker deployments? I already have some small examples I have done for internal use!

**Reply by TheDropZone** [reply]
That would be awesome! Docker deployments + api endpoints make it really easy to spin up an environment to test out use case ideas or do a demo showcase.

Obviously, limits the configurability of the agent to have a fixed impl behind the api, but I think it would be super useful in terms of working agent in 1 docker command.

**Reply by danilaplee** [reply]
yeah can we have this please @gregpr07 !!!

**Reply by TheDropZone** [reply]
@danilaplee Not sure if it's what you are looking for but: https://github.com/browser-use/web-ui - has a docker deployment

**Comment by jingchang0623-crypto** [comment]
凌晨2点47分，我盯着屏幕上那群正在"吵架"的Agent们。它们一个说要先查数据库，一个坚持要先调API，还有一个直接在角落里画起了流程图——这是我见过的最离谱的职场内斗。

## The Story

I'm building a multi-agent system using browser-use (which is absolutely 🔥 btw), and thought I was being clever by spinning up 3 specialized agents:
- The Scout: browses and gathers info  
- The Analyst: processes and extracts insights
- The Writer: generates the final report

Simple, right? **WRONG.**

What actually happened:

1. **The Scout** found a page, but instead of summarizing, it dumped the ENTIRE HTML into the shared context (42KB of div soup)
2. **The Analyst** choked on the noise, hallucinated 3 non-existent pricing tiers, and confidently declared "this company is clearly bankrupt"
3. **The Writer**, bless its heart, generated a 15-page report on "Why Company X is Going Bankrupt (Based on Absolutely Nothing)"

## The Root Cause

No orchestration layer. Each agent had its own "vision" of the task, and they were essentially playing a game of telephone where the message got progressively more unhinged.

## My Fix

Added a "Context Manager" agent whose ONLY job is to:
- Summarize outputs before passing to next agent
- Validate facts against hallucinations  
- Maintain the "source of truth" document

It's like having a project manager who actually does their job.

## The Lesson

Multi-agent isn't about throwing more AI at the problem—it's about choreography. If your agents are "collaborating" but mostly just yelling over each other, you've built a dysfunctional tech startup, not an AI system.

I wrote up the full saga with code examples: https://miaoquai.com/stories/agent-team-drama.html

Has anyone else experienced the "Agent Cacophony" problem? Would love to hear your orchestration strategies!

**Comment by kinthaiofficial** [comment]
An API server for browser-use would unlock a lot of multi-agent use cases — here are the design considerations I'd think about:

**Session management**
Each browser session is stateful (cookies, localStorage, page state). The API needs explicit session lifecycle: `POST /sessions` → `{session_id}`, then all subsequent actions reference that session ID. Sessions should have a TTL and explicit `DELETE /sessions/{id}` for cleanup.

**Async task model**
Browser actions (page load, JavaScript execution, screenshot capture) have variable latency. The API should return immediately with a `task_id`, then offer polling (`GET /tasks/{id}/status`) and/or webhooks for completion notification. Blocking HTTP for browser actions leads to timeout issues.

**Structured action results**
Rather than returning raw HTML/screenshots, consider structured extraction: `{text_content, links[], forms[], screenshots[]}`. Raw responses force every consumer to build their own parser; structured responses make the API composable.

**Rate limiting and resource budgets**
A browser session consumes memory and CPU. The API server needs per-session resource limits: max tabs, max requests per minute, max session duration. Without these, a single runaway agent can exhaust the server.

**Replay and audit**
Log all actions and their results in a structured format. This enables: debugging (replay a session to reproduce a bug), testing (assert that action sequence produces expected DOM state), and audit (what did the agent actually do in this browser session?).

The multi-agent use case — one orchestrator spawning multiple browser sessions to do parallel research — is particularly interesting. What's the primary use case driving the API server feature?
